integration/container/kill_test.go

```diff
@@ -0,0 +1,182 @@
+	require.NoError(t, err)
+
+	err = client.ContainerKill(ctx, c.ID, "0")
+	require.EqualError(t, err, "Error response from daemon: Invalid signal: 0")
+	poll.WaitOn(t, containerIsInState(ctx, client, c.ID, "running"), poll.WithDelay(100*time.Millisecond))
+
+	err = client.ContainerKill(ctx, c.ID, "SIG42")
+	require.EqualError(t, err, "Error response from daemon: Invalid signal: SIG42")
+	poll.WaitOn(t, containerIsInState(ctx, client, c.ID, "running"), poll.WithDelay(100*time.Millisecond))
+}
+
+func TestKillContainer(t *testing.T) {
+	defer setupTest(t)()
+	client := request.NewAPIClient(t)
+
+	testCases := []struct {
+		doc    string
+		signal string
+		status string
+	}{
+		{
+			doc:    "no signal",
+			signal: "",
+			status: "exited",
+		},
+		{
+			doc:    "non killing signal",
+			signal: "SIGWINCH",
+			status: "running",
+		},
+		{
+			doc:    "killing signal",
+			signal: "SIGTERM",
+			status: "exited",
+		},
+	}
+
+	for _, tc := range testCases {
+		tc := tc
+		t.Run(tc.doc, func(t *testing.T) {
+			t.Parallel()
+			ctx := context.Background()
+			c, err := client.ContainerCreate(ctx,
+				&container.Config{
+					Image: "busybox",
+					Cmd:   strslice.StrSlice([]string{"top"}),
+				},
+				&container.HostConfig{},
+				&network.NetworkingConfig{},
+				"")
+			require.NoError(t, err)
+			err = client.ContainerStart(ctx, c.ID, types.ContainerStartOptions{})
+			require.NoError(t, err)
+			err = client.ContainerKill(ctx, c.ID, tc.signal)
+			require.NoError(t, err)
+
+			poll.WaitOn(t, containerIsInState(ctx, client, c.ID, tc.status), poll.WithDelay(100*time.Millisecond))
+		})
+	}
+}
+
+func TestKillWithStopSignalAndRestartPolicies(t *testing.T) {
+	skip.If(t, testEnv.OSType != "linux", "Windows only supports 1.25 or later")
+	defer setupTest(t)()
+	client := request.NewAPIClient(t)
+
+	testCases := []struct {
+		doc        string
+		stopsignal string
+		status     string
+	}{
+		{
+			doc:        "same-signal-disables-restart-policy",
+			stopsignal: "TERM",
+			status:     "exited",
+		},
+		{
+			doc:        "different-signal-keep-restart-policy",
+			stopsignal: "CONT",
+			status:     "running",
+		},
+	}
+
+	for _, tc := range testCases {
+		tc := tc
+		t.Run(tc.doc, func(t *testing.T) {
+			t.Parallel()
+			ctx := context.Background()
+			c, err := client.ContainerCreate(ctx,
+				&container.Config{
+					Image:      "busybox",
+					Cmd:        strslice.StrSlice([]string{"top"}),
+					StopSignal: tc.stopsignal,
+				},
+				&container.HostConfig{
+					RestartPolicy: container.RestartPolicy{
+						Name: "always",
+					}},
+				&network.NetworkingConfig{},
+				"")
+			require.NoError(t, err)
+			err = client.ContainerStart(ctx, c.ID, types.ContainerStartOptions{})
+			require.NoError(t, err)
+			err = client.ContainerKill(ctx, c.ID, "TERM")
+			require.NoError(t, err)
+
+			poll.WaitOn(t, containerIsInState(ctx, client, c.ID, tc.status), poll.WithDelay(100*time.Millisecond))
```

Comment:
Just a performance concern here.. could this start up a "wait" goroutine earlier on for the case of it being stopped? (side thought, would be nice if the API let us wait for a container to be running).

Reply:
Funny, just found https://github.com/moby/moby/pull/36147